cpp/arcticdb/util/decimal.cpp

```diff
@@ -0,0 +1,328 @@
+            arcticdb::util::check_arg(
+                    size_ + zeros_to_append <= max_digits,
+                    "Cannot parse decimal from string. "
+                    "Overflow. Input has more than {} significant digits.",
+                    max_digits);
+            for(int i = 0; i < zeros_to_append; ++i) {
+                digits_[size_++] = '0';
+            }
+        }
+
+        bool NumberComponents::is_negative() const {
+            return flags_ & NumberComponentsFlags::NEGATIVE;
+        }
+
+        bool NumberComponents::is_decimal() const {
+            return flags_ & NumberComponentsFlags::DECIMAL;
+        }
+
+        const char* NumberComponents::get_digits() const {
+            return digits_.data();
+        }
+
+        int NumberComponents::get_size() const {
+            return size_;
+        }
+
+        Decimal::Decimal() : data_{ 0 } {}
+
+        Decimal::Decimal(std::string_view number) : data_{0} {
+            // GCC and Clang have internal 2's complement __uint128_t.
+            // MSVC does not have 128-bit integer, it has __m128, which is for SIMD.
+            // Boost's uint128_t is not in two's complement, so it cannot be reinterpret_cast over the data.
+            // TODO: potential optimization for Clang/GCC would be to load it __uint128_t and reinterpret_cast
+            //  it over the array.
+            const NumberComponents components(number);
+            std::string_view number_to_parse(components.get_digits());
+            while(!number_to_parse.empty()) {
+                const std::string_view chunk = number_to_parse.substr(0, max_digits_per_word);
+                push_chunk(chunk);
+                number_to_parse = number_to_parse.substr(chunk.size());
+            }
+            if(components.is_negative()) {
+                negate();
+            }
+        }
+
+        void Decimal::push_chunk(std::string_view chunk_str) {
+            uint64_t chunk = to_uint64_t(chunk_str);
+            const uint64_t word_exponent = powers_of_10[chunk_str.length()];
+            const uint64_t mask = 0xFFFFFFFFFFFFFFFFULL;
+            for(uint64_t &word : data_) {
+                boost::multiprecision::uint128_t tmp = word;
+                tmp *= word_exponent;
+                tmp += chunk;
+                word = static_cast<uint64_t>(tmp & mask);
+                chunk = static_cast<uint64_t>(tmp >> 64);
+            }
+        }
+
+        std::string Decimal::to_string(int scale) const {
+            if(is_negative()) {
+                const Decimal negated = this->operator-();
+                std::string result = negated.to_string(scale);
+                result.insert(0, "-");
+                return result;
+            }
+            if(data_[0] == 0 && data_[1] == 0) {
+                return "0";
+            }
+
+            boost::multiprecision::uint128_t number = to_uint128(
```

Comment:
This also uses boost's implementation as intermediate type. Again we could change this if needed at the cost of having a bit more complicated code.